### ISSUE TITLE
API Changes - 0.4

### ISSUE THREAD

**Issue by rrag** [rrag/react-stockcharts#48]
This issue will serve as a complete list of all the **breaking** API changes I am thinking about. Get feedback from users. Based on feedback and progress, this will be updated to keep current

### v0.4

The below changes are aimed at simplifying the internals, and provide a way to have indicators which depend on other indicators.

1. add a new prop `calculate={[ Calculator1, Calculator2 ]}` to the `ChartCanvas`, like so

        <ChartCanvas width={...} height={...} ...
            calculator={[ { EMACalculator, MACDCalculator, RSICalculator } ]}
where

        EMACalculator = ema()
          .window(12) // optional will default to 10
          .source(d => d.close) // optional will default to close as the source
          .merge((d, c) => {d.ema12 = c}) // Required, if not provided, log a error
          .yAccessor(d => d.ema12) // Required, if not provided, log an error during calculation
          .stroke("blue") // Optional

        MACDCalculator = macd()
          .fastWindow(...) // optional will default to 12
          .slowWindow(...) // optional will default to 26
          .signalWindow(...) // optional will default to 9
          .source(d => d.close) // optional will default to close as the source
          .merge((d, c) => {d.macd = c});

  `ema`, `macd` above will be provided by the library

1. add a prop `yExtents` to `Chart` which will take an array of functions, like so

        <Chart id={...} ... yExtents={[ d => ([d.high, d.low]), d => d.ema12, d => d.sma100 ]}
  Right now, the way the upper and lower bounds are calculated is by taking all the `yAccessor`s of the `DataSeries` and finding the max and min values of them. `yExtents` will serve that purpose. If you want to fix the y domain between say 20 and 100 it can be achieved by `yExtents={[ 20, 100 ]}`

1. Remove `DataSeries`, Since the purpose of `DataSeries` is to provide a `yAccessor` or specify an `indicator`, and both these functions are moved to other components, this can be removed

1. All Series will take a mandatory `yAccessor` property. like so

        <LineSeries yAccessor={d => d.ema12}  />
  most of the series can still define a default `yAccessor` e.g. `CandlesticSeries` will have the ohlc as the `yAccessor`

1. `CurrentCoordinate` and `EdgeIndicator` will now take a `yAccessor` instead of `forChart={...} forDataSeries={...}`. Since the `forChart` and `forDataSeries` props will be removed, this also means that the `stroke` and `fill` attributed cannot be auto derived by the component, but it can still be done

        // Before
        <CurrentCoordinate forChart={2} forDataSeries={0} />

        // After
        <CurrentCoordinate yAccessor={EMACalculator.yAccessor()} stroke={EMACalculator.stroke()} />

1. Tooltips will also take a `yAccessor`, Since many of the indicators, have their own tooltips, we can even do

        // Before
        <MACDTooltip forChart={3} origin={[-38, 15]}/>

        // After
        <MACDTooltip origin={[-38, 15]} calculator={MACDCalculator} />

1. With this change, `dataTransform` and `calculator` props on `<ChartCanvas ... >` seem redundant, they can be combined. so we can drop `dataTransform`

1. Might as well move `CurrentCoordinate` and `EdgeContainer` `EdgeIndicator` inside the `Chart`. Have to see if this will bring any performance issues. This way we no longer refer to a chart and series by numbers

1. add props `xAccessor={d => d.date}` `xExtents={[ leftDate, rightDate ]}` `xScale={stockscale}` to `ChartCanvas` this will 

EDIT: Added removal of `dataTransform` prop and moving `CurrentCoordinate` and `EdgeContainer` `EdgeIndicator` inside `Chart`

EDIT: add props `xAccessor`, `xExtents` and `xScale` to `ChartCanvas`

**Comment by rrag** [comment]
Before
```jsx
<ChartCanvas ref="chartCanvas" width={width} height={600}
	margin={{left: 70, right: 70, top:20, bottom: 30}} initialDisplay={200} 
	dataTransform={[ { transform: StockscaleTransformer } ]}
	data={data} type={type}>
	<Chart id={1} yMousePointerDisplayLocation="right" height={400}
			yMousePointerDisplayFormat={d3.format(".2f")} padding={{ top: 10, right: 0, bottom: 20, left: 0 }}>
		<YAxis axisAt="right" orient="right" ticks={5} />
		<XAxis axisAt="bottom" orient="bottom" showTicks={false} outerTickSize={0} />
		<DataSeries id={0} yAccessor={CandlestickSeries.yAccessor} >
			<CandlestickSeries />
		</DataSeries>
		<DataSeries id={1} indicator={EMA} options={{ period: 26 }} >
			<LineSeries/>
		</DataSeries>
		<DataSeries id={2} indicator={EMA} options={{ period: 12 }} >
			<LineSeries/>
		</DataSeries>
	</Chart>
	<CurrentCoordinate forChart={1} forDataSeries={1} />
	<CurrentCoordinate forChart={1} forDataSeries={2} />
	<Chart id={2} yMousePointerDisplayLocation="left" yMousePointerDisplayFormat={d3.format(".4s")}
			height={150} origin={(w, h) => [0, h - 300]} >
		<YAxis axisAt="left" orient="left" ticks={5} tickFormat={d3.format("s")}/>
		<DataSeries id={0} yAccessor={(d) => d.volume} >
			<HistogramSeries
				fill={(d) => d.close > d.open ? "#6BA583" : "#FF0000"}
				opacity={0.5} />
		</DataSeries>
		<DataSeries id={1} indicator={SMA} options={{ period: 10, source:"volume" }} stroke="#4682B4" fill="#4682B4">
			<AreaSeries opacity={0.5} />
		</DataSeries>
	</Chart>
	<CurrentCoordinate forChart={2} forDataSeries={0} />
	<CurrentCoordinate forChart={2} forDataSeries={1} />
	<EdgeContainer>
		<EdgeIndicator itemType="last" orient="right"
			edgeAt="right" forChart={1} forDataSeries={1} />
		<EdgeIndicator itemType="last" orient="right"
			edgeAt="right" forChart={1} forDataSeries={2} />
		<EdgeIndicator itemType="first" orient="left"
			edgeAt="left" forChart={1} forDataSeries={1} />
		<EdgeIndicator itemType="first" orient="left"
			edgeAt="left" forChart={1} forDataSeries={2} />
	</EdgeContainer>
	<Chart id={3} yMousePointerDisplayLocation="right" yMousePointerDisplayFormat={d3.format(".2f")}
			height={150} origin={(w, h) => [0, h - 150]} padding={{ top: 10, right: 0, bottom: 10, left: 0 }} >
		<XAxis axisAt="bottom" orient="bottom"/>
		<YAxis axisAt="right" orient="right" ticks={2}/>
		<DataSeries id={0} indicator={MACD} options={{ fast: 12, slow: 26, signal: 9 }} >
			<MACDSeries />
		</DataSeries>
	</Chart>
	<MouseCoordinates xDisplayFormat={d3.time.format("%Y-%m-%d")} />
	<EventCapture mouseMove={true} zoom={true} pan={true} mainChart={1} defaultFocus={false} />
	<TooltipContainer>
		<OHLCTooltip forChart={1} origin={[-40, -10]}/>
		<MovingAverageTooltip forChart={1} onClick={(e) => console.log(e)} origin={[-38, 5]} />
		<MACDTooltip forChart={3} origin={[-38, 15]}/>
	</TooltipContainer>
</ChartCanvas>
```

After

```jsx
<ChartCanvas ref="chartCanvas" width={width} height={600}
	margin={{left: 70, right: 70, top:20, bottom: 30}} initialDisplay={200} 
	dataTransform={[ { transform: StockscaleTransformer } ]}
	calculator={[EMA26, EMA12, MACD, VolumeSMA10]}
	data={data} type={type}>
	<Chart id={1} yMousePointerDisplayLocation="right" height={400}
			yMousePointerDisplayFormat={d3.format(".2f")} padding={{ top: 10, right: 0, bottom: 20, left: 0 }}
			yExtents={[d => ([d.high, d.low]), EMA26.yAccessor(), EMA12.yAccessor() ]}>
		<YAxis axisAt="right" orient="right" ticks={5} />
		<XAxis axisAt="bottom" orient="bottom" showTicks={false} outerTickSize={0} />
		<CandlestickSeries />
		<LineSeries yAccessor={EMA26.yAccessor()} stroke={EMA26.stroke()} />
		<LineSeries yAccessor={EMA12.yAccessor()} stroke={EMA12.stroke()} />
	</Chart>
	<CurrentCoordinate yAccessor={EMA26.yAccessor()} stroke={EMA26.stroke()} />
	<CurrentCoordinate yAccessor={EMA12.yAccessor()} stroke={EMA12.stroke()} />
	<Chart id={2} yMousePointerDisplayLocation="left" yMousePointerDisplayFormat={d3.format(".4s")}
			height={150} origin={(w, h) => [0, h - 300]}
			yExtents={[d => d.volume, VolumeSMA10.yAccessor()]}>
		<YAxis axisAt="left" orient="left" ticks={5} tickFormat={d3.format("s")}/>
		<HistogramSeries
			yAccessor={(d) => d.volume}
			fill={(d) => d.close > d.open ? "#6BA583" : "#FF0000"}
			opacity={0.5} />
		<AreaSeries opacity={0.5} yAccessor={VolumeSMA10.yAccessor()} stroke={VolumeSMA10.stroke()} fill={VolumeSMA10.stroke()}/>
	</Chart>
	<CurrentCoordinate yAccessor={d => d.volume} stroke="#000000" />
	<CurrentCoordinate yAccessor={VolumeSMA10.yAccessor()} stroke={VolumeSMA10.stroke()} />
	<EdgeContainer>
		<EdgeIndicator itemType="last" orient="right"
			edgeAt="right" yAccessor={EMA26.yAccessor()} fill={EMA26.stroke()} />
		<EdgeIndicator itemType="last" orient="right"
			edgeAt="right" yAccessor={EMA12.yAccessor()} fill={EMA12.stroke()} />
		<EdgeIndicator itemType="first" orient="left"
			edgeAt="left" yAccessor={EMA26.yAccessor()} fill={EMA26.stroke()} />
		<EdgeIndicator itemType="first" orient="left"
			edgeAt="left" yAccessor={EMA12.yAccessor()} fill={EMA12.stroke()} />
	</EdgeContainer>
	<Chart id={3} yMousePointerDisplayLocation="right" yMousePointerDisplayFormat={d3.format(".2f")}
			height={150} origin={(w, h) => [0, h - 150]} padding={{ top: 10, right: 0, bottom: 10, left: 0 }}
			yExtents={[MACD.yAccessor()]}>
		<XAxis axisAt="bottom" orient="bottom"/>
		<YAxis axisAt="right" orient="right" ticks={2}/>
		<MACDSeries />
	</Chart>
	<MouseCoordinates xDisplayFormat={d3.time.format("%Y-%m-%d")} />
	<EventCapture mouseMove={true} zoom={true} pan={true} mainChart={1} defaultFocus={false} />
	<TooltipContainer>
		<OHLCTooltip forChart={1} origin={[-40, -10]}/>
		<MovingAverageTooltip forChart={1} origin={[-38, 5]} yAccessors={[EMA12.yAccessor(), EMA26.yAccessor()}]} />
		<MACDTooltip forChart={3} origin={[-38, 15]} calculator={MACD}/>
	</TooltipContainer>
</ChartCanvas>
```

**Comment by rrag** [comment]
1, 2, 3, 5, 7, 8, 9 - Done

4, 6 - Work in progress

Other issues:
[ ] Window resize is not updating the chart
[ ] Zooming from one interval to next is jumpy, and not smooth
[ ] touch zoom and pan is broken
[ ] compare is broken
[ ] remove unused code in utils.js, review other files too.

**Comment by rrag** [comment]
released `0.4.0-beta2`

This will most likely go out as `0.4.0` in a couple of days
